File: assets/Scripts/Modules/MagicTiles/TapValidator.cs
using System;
using UnityEngine;

namespace Game.MagicTiles
{
    /// <summary>
    /// Validates tap accuracy based on timing and tile position.
    /// Determines if a tap is a hit or miss and calculates the quality of the hit.
    /// </summary>
    public class TapValidator : MonoBehaviour
    {
        [Header("Validation Settings")]
        [SerializeField] private float perfectHitWindow = 0.05f; // Time window for perfect hit in seconds
        [SerializeField] private float goodHitWindow = 0.15f; // Time window for good hit in seconds
        [SerializeField] private bool allowTapsWithoutTiles = false; // Whether to allow taps with no tiles
        
        [Header("Combo Settings")]
        [SerializeField] private int comboBreakOnMiss = true; // Whether missing breaks combo
        [SerializeField] private int minimumComboForMultiplier = 4; // Min combo before multiplier kicks in
        
        // References to other managers
        private TileManager tileManager;
        private AudioManager audioManager;
        private BeatmapManager beatmapManager;
        private GameplayManager gameplayManager;
        
        // Combo tracking
        private int currentCombo = 0;
        private int maxCombo = 0;
        
        // Events
        public event Action<HitResult> OnHitResultDetermined;
        public event Action<int> OnComboChanged;
        public event Action<int> OnMaxComboChanged;
        
        private void Awake()
        {
            // Get references to other managers
            tileManager = FindObjectOfType<TileManager>();
            audioManager = FindObjectOfType<AudioManager>();
            beatmapManager = FindObjectOfType<BeatmapManager>();
            gameplayManager = FindObjectOfType<GameplayManager>();
        }
        
        /// <summary>
        /// Validates a tap in a specific lane
        /// </summary>
        /// <param name="lane">Lane that was tapped</param>
        /// <param name="position">Position of the tap</param>
        public void ValidateTap(int lane, Vector2 position)
        {
            // Make sure the game is playing
            if (gameplayManager != null && gameplayManager.CurrentGameState != GameStates.Playing)
                return;
                
            // Get the current music position
            float currentTime = audioManager != null ? audioManager.GetMusicPosition() : Time.time;
            
            // Try to get a tile at hit position
            Tile tile = tileManager.GetTileAtHitPosition(lane);
            
            if (tile != null)
            {
                // Tile exists at hit position, validate against beatmap timing
                ValidateAgainstBeatmap(lane, currentTime, tile);
            }
            else if (allowTapsWithoutTiles)
            {
                // Allow for tapping without tiles (might be useful for tutorial or certain game modes)
                // But reset combo
                ResetCombo();
                
                // Trigger miss effect
                OnHitResultDetermined?.Invoke(new HitResult(false, HitQuality.Miss, 0f, null));
                
                // Play miss sound
                if (audioManager != null)
                {
                    audioManager.PlaySoundEffect(AudioManager.SoundEffectType.Miss);
                }
            }
        }
        
        /// <summary>
        /// Validates a tap against the beatmap data
        /// </summary>
        /// <param name="lane">Lane that was tapped</param>
        /// <param name="currentTime">Current time in the song</param>
        /// <param name="tile">Tile that was tapped</param>
        private void ValidateAgainstBeatmap(int lane, float currentTime, Tile tile)
        {
            // Get beat data from tile
            BeatData beatData = tile.GetBeatData();
            
            // Calculate timing difference
            float timeDifference = Mathf.Abs(currentTime - beatData.Time);
            
            // Determine hit quality
            HitQuality quality = HitQuality.Miss;
            bool isHit = false;
            
            if (timeDifference <= perfectHitWindow)
            {
                quality = HitQuality.Perfect;
                isHit = true;
            }
            else if (timeDifference <= goodHitWindow)
            {
                quality = HitQuality.Good;
                isHit = true;
            }
            
            // Create hit result
            HitResult result = new HitResult(isHit, quality, timeDifference, beatData);
            
            // Update combo
            if (isHit)
            {
                IncrementCombo();
                
                // Play hit effect on tile
                tileManager.PlayHitEffect(tile, quality);
                
                // Play sound effect
                if (audioManager != null)
                {
                    if (quality == HitQuality.Perfect)
                    {
                        audioManager.PlaySoundEffect(AudioManager.SoundEffectType.PerfectTap);
                    }
                    else
                    {
                        audioManager.PlaySoundEffect(AudioManager.SoundEffectType.Tap);
                    }
                }
            }
            else
            {
                // Miss
                if (comboBreakOnMiss)
                {
                    ResetCombo();
                }
                
                // Play miss effect on tile
                tileManager.PlayMissEffect(tile);
                
                // Play miss sound
                if (audioManager != null)
                {
                    audioManager.PlaySoundEffect(AudioManager.SoundEffectType.Miss);
                }
            }
            
            // Notify listeners of hit result
            OnHitResultDetermined?.Invoke(result);
        }
        
        /// <summary>
        /// Called when a tile reaches the end without being tapped
        /// </summary>
        /// <param name="tile">Tile that was missed</param>
        public void OnTileMissed(Tile tile)
        {
            // Reset combo on miss
            if (comboBreakOnMiss)
            {
                ResetCombo();
            }
            
            // Create miss result
            HitResult result = new HitResult(false, HitQuality.Miss, 0f, tile.GetBeatData());
            
            // Play miss effect
            tileManager.PlayMissEffect(tile);
            
            // Play miss sound
            if (audioManager != null)
            {
                audioManager.PlaySoundEffect(AudioManager.SoundEffectType.Miss);
            }
            
            // Notify listeners of miss
            OnHitResultDetermined?.Invoke(result);
        }
        
        /// <summary>
        /// Increments the current combo
        /// </summary>
        private void IncrementCombo()
        {
            currentCombo++;
            if (currentCombo > maxCombo)
            {
                maxCombo = currentCombo;
                OnMaxComboChanged?.Invoke(maxCombo);
            }
            
            OnComboChanged?.Invoke(currentCombo);
        }
        
        /// <summary>
        /// Resets the current combo to zero
        /// </summary>
        private void ResetCombo()
        {
            if (currentCombo > 0)
            {
                currentCombo = 0;
                OnComboChanged?.Invoke(currentCombo);
            }
        }
        
        /// <summary>
        /// Gets the current combo
        /// </summary>
        /// <returns>Current combo count</returns>
        public int GetCombo()
        {
            return currentCombo;
        }
        
        /// <summary>
        /// Gets the maximum combo achieved
        /// </summary>
        /// <returns>Maximum combo</returns>
        public int GetMaxCombo()
        {
            return maxCombo;
        }
        
        /// <summary>
        /// Calculates the score multiplier based on current combo
        /// </summary>
        /// <returns>Score multiplier (1.0 or higher)</returns>
        public float GetComboMultiplier()
        {
            if (currentCombo < minimumComboForMultiplier)
                return 1.0f;
                
            // Calculate multiplier, maxing out at 2x
            float multiplier = 1.0f + Mathf.Min(1.0f, (currentCombo - minimumComboForMultiplier) / 50.0f);
            return multiplier;
        }
        
        /// <summary>
        /// Resets the validator for a new game
        /// </summary>
        public void Reset()
        {
            currentCombo = 0;
            maxCombo = 0;
            OnComboChanged?.Invoke(currentCombo);
            OnMaxComboChanged?.Invoke(maxCombo);
        }
    }
} 

File: assets/Scripts/Modules/MagicTiles/TapValidator.ts
import { _decorator, Component, Node } from 'cc';
import { HitRating } from './Tile';
import { BeatmapManager } from './BeatmapManager';

const { ccclass, property } = _decorator;

/**
 * Interface for combo data
 */
interface ComboData {
    count: number;
    maxCount: number;
    lastRating: HitRating;
    perfectCount: number;
    goodCount: number;
    okCount: number;
    missCount: number;
}

/**
 * TapValidator for Magic Tiles 3
 * Validates tap timing accuracy and manages combos
 */
@ccclass('TapValidator')
export class TapValidator extends Component {
    // Audio offset (milliseconds) - to adjust for audio latency
    @property
    audioOffset: number = 0;
    
    // Perfect hit window (seconds)
    @property
    perfectWindow: number = 0.05; // 50ms
    
    // Good hit window (seconds)
    @property
    goodWindow: number = 0.1; // 100ms
    
    // OK hit window (seconds)
    @property
    okWindow: number = 0.15; // 150ms
    
    // Reference to beatmap manager
    private beatmapManager: BeatmapManager = null!;
    
    // Combo data
    private combo: ComboData = {
        count: 0,
        maxCount: 0,
        lastRating: HitRating.MISS,
        perfectCount: 0,
        goodCount: 0,
        okCount: 0,
        missCount: 0
    };
    
    // Callback for combo change events
    private onComboChangeCallback: ((combo: number) => void) | null = null;
    
    // Callback for rating events
    private onRatingCallback: ((lane: number, rating: HitRating) => void) | null = null;
    
    onLoad() {
        this.resetCombo();
        this.beatmapManager = BeatmapManager.instance;
    }
    
    /**
     * Validate a tap at a specific time
     * @param lane The lane that was tapped
     * @param tapTime The time of the tap
     * @param rating The hit rating from the tile manager
     * @returns The hit rating for the tap
     */
    validateTap(lane: number, tapTime: number, rating: HitRating): HitRating {
        // Update combo based on rating
        this.updateCombo(rating);
        
        // Notify listeners of the rating
        if (this.onRatingCallback) {
            this.onRatingCallback(lane, rating);
        }
        
        return rating;
    }
    
    /**
     * Calculate timing accuracy for a tap
     * @param tapTime The time of the tap
     * @param expectedTime The expected time for the tap
     * @returns The hit rating based on timing accuracy
     */
    calculateAccuracy(tapTime: number, expectedTime: number): HitRating {
        // Apply audio offset
        const adjustedTapTime = tapTime + (this.audioOffset / 1000);
        
        // Calculate timing difference
        const timeDiff = Math.abs(adjustedTapTime - expectedTime);
        
        // Determine rating based on timing windows
        if (timeDiff <= this.perfectWindow) {
            return HitRating.PERFECT;
        } else if (timeDiff <= this.goodWindow) {
            return HitRating.GREAT;
        } else if (timeDiff <= this.okWindow) {
            return HitRating.COOL;
        } else {
            return HitRating.MISS;
        }
    }
    
    /**
     * Update the combo based on the hit rating
     * @param rating The hit rating
     */
    private updateCombo(rating: HitRating) {
        // Update rating counts
        switch (rating) {
            case HitRating.PERFECT:
                this.combo.perfectCount++;
                this.combo.count++;
                break;
            case HitRating.GREAT:
                this.combo.goodCount++;
                this.combo.count++;
                break;
            case HitRating.COOL:
                this.combo.okCount++;
                this.combo.count++;
                break;
            case HitRating.MISS:
                this.combo.missCount++;
                this.combo.count = 0; // Reset combo on miss
                break;
        }
        
        // Update max combo
        if (this.combo.count > this.combo.maxCount) {
            this.combo.maxCount = this.combo.count;
        }
        
        // Store last rating
        this.combo.lastRating = rating;
        
        // Notify combo change listeners
        if (this.onComboChangeCallback) {
            this.onComboChangeCallback(this.combo.count);
        }
    }
    
    /**
     * Reset the combo and statistics
     */
    resetCombo() {
        this.combo = {
            count: 0,
            maxCount: 0,
            lastRating: HitRating.MISS,
            perfectCount: 0,
            goodCount: 0,
            okCount: 0,
            missCount: 0
        };
        
        // Notify combo change listeners
        if (this.onComboChangeCallback) {
            this.onComboChangeCallback(0);
        }
    }
    
    /**
     * Get the current combo
     */
    getCombo(): number {
        return this.combo.count;
    }
    
    /**
     * Get the max combo
     */
    getMaxCombo(): number {
        return this.combo.maxCount;
    }
    
    /**
     * Get the accuracy percentage
     * @returns Accuracy as a percentage (0-100)
     */
    getAccuracy(): number {
        const totalNotes = this.combo.perfectCount + this.combo.goodCount + this.combo.okCount + this.combo.missCount;
        if (totalNotes === 0) return 100; // No notes hit yet
        
        // Weight different ratings
        const weightedSum = (this.combo.perfectCount * 100) + (this.combo.goodCount * 80) + (this.combo.okCount * 50);
        return Math.round((weightedSum / (totalNotes * 100)) * 100);
    }
    
    /**
     * Get the counts of each rating
     */
    getRatingCounts(): { perfect: number, good: number, ok: number, miss: number } {
        return {
            perfect: this.combo.perfectCount,
            good: this.combo.goodCount,
            ok: this.combo.okCount,
            miss: this.combo.missCount
        };
    }
    
    /**
     * Get the total number of notes hit
     */
    getTotalNotes(): number {
        return this.combo.perfectCount + this.combo.goodCount + this.combo.okCount + this.combo.missCount;
    }
    
    /**
     * Register a callback for combo changes
     * @param callback Function to call when the combo changes
     */
    onComboChange(callback: (combo: number) => void) {
        this.onComboChangeCallback = callback;
    }
    
    /**
     * Register a callback for rating events
     * @param callback Function to call when a rating is determined
     */
    onRating(callback: (lane: number, rating: HitRating) => void) {
        this.onRatingCallback = callback;
    }
    
    /**
     * Set the audio offset
     * @param offsetMs Offset in milliseconds
     */
    setAudioOffset(offsetMs: number) {
        this.audioOffset = offsetMs;
    }
    
    /**
     * Get the audio offset
     * @returns Offset in milliseconds
     */
    getAudioOffset(): number {
        return this.audioOffset;
    }
    
    /**
     * Set the hit windows
     * @param perfect Perfect hit window in seconds
     * @param good Good hit window in seconds
     * @param ok OK hit window in seconds
     */
    setHitWindows(perfect: number, good: number, ok: number) {
        this.perfectWindow = perfect;
        this.goodWindow = good;
        this.okWindow = ok;
    }
    
    /**
     * Check if a combo milestone has been reached
     * @param milestone The milestone to check (e.g., 50, 100, 200)
     * @returns Whether the milestone was just reached
     */
    isComboMilestone(milestone: number): boolean {
        return this.combo.count === milestone;
    }
    
    /**
     * Calculate a score based on performance
     * @returns Score value
     */
    calculateScore(): number {
        const comboBonus = Math.min(this.combo.maxCount / 10, 100); // Cap at 100
        const accuracyBonus = this.getAccuracy();
        
        // Calculate base score from rating counts
        const baseScore = (this.combo.perfectCount * 1000) + 
                         (this.combo.goodCount * 500) + 
                         (this.combo.okCount * 100);
        
        // Apply bonuses
        return Math.round(baseScore * (1 + (comboBonus / 100) + (accuracyBonus / 100)));
    }
} 